### ISSUE TITLE
Possible cast_possible_truncation false positive

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#8316]
### Summary

This code:

```rust
#![warn(clippy::pedantic)]

fn main() {
    const X: usize = 100_000;
    const X2: u32 = X as _;
    dbg!(X);
    dbg!(X2);
}
```

Gives:

```rust
warning: casting `usize` to `u32` may truncate the value on targets with 64-bit wide pointers
 --> src/main.rs:5:21
  |
5 |     const X2: u32 = X as _;
  |                     ^^^^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::pedantic)]
  |         ^^^^^^^^^^^^^^^^
  = note: `#[warn(clippy::cast_possible_truncation)]` implied by `#[warn(clippy::pedantic)]`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_possible_truncation
```

Perhaps this pedantic warning could be generated only if there's an actual truncation. Here 100_000 fits in 32 bits.

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

**Comment by jyn514** [comment]
I think this is a duplicate of https://github.com/rust-lang/rust-clippy/issues/9613.

**Comment by leonardo-m** [comment]
Right.
